### ISSUE TITLE
Trimming Error When Publishing Project Using Source Generator

### ISSUE THREAD

**Issue by ritchiecarroll** [dotnet/roslyn#76132]
## Issue Summary

When publishing a project that uses a source generator with trimming enabled, the build fails with an error suggesting .NET Core 3.0+ is required, despite the project targeting .NET 9.0 and the source generator being marked as non-referenced build-time component.

## Version Used

Visual Studio 17.12.0 / .NET 9.0 / C# 13

## Steps to Reproduce

1. Create a .NET Standard 2.0 source generator project (as required for analyzers)
2. Create a .NET 9.0 project that references this source generator, using code that depends on the generated output
3. Reference the generator in the consuming project like so (normally):
   ```xml
   <ProjectReference Include="..\MyGenerator.csproj" 
                     OutputItemType="Analyzer" 
                     ReferenceOutputAssembly="false" 
                     PrivateAssets="All" />
   ```
4. Build project normally
5. Attempt to _publish_ the consuming project to "Folder" with trimming enabled

## Expected Behavior

The project should publish successfully with trimming enabled. Since the source generator is only used at build time and is marked with `ReferenceOutputAssembly="false"`, it should not be considered during the trimming process.

## Actual Behavior

Build fails with error `NETSDK1124`:
```
C:\Program Files\dotnet\sdk\9.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(90,5):
Error NETSDK1124: Trimming assemblies requires .NET Core 3.0 or higher.
```

The error occurs despite:
1. The consuming project targeting .NET 9.0
2. The analyzer being marked as not referenced in the output
3. The analyzer being a build-time-only component

## Workaround

The issue can be worked around by conditionally by first building project normally, emitting generated files into a local folder. Folder needs to be excluded for normal builds, but available when publishing trimmed, then also excluding the source generator during trimmed publish. For example:

```xml
  <PropertyGroup>
    <EmitCompilerGeneratedFiles>true</EmitCompilerGeneratedFiles>
    <CompilerGeneratedFilesOutputPath>generated</CompilerGeneratedFilesOutputPath>
  </PropertyGroup>

  <ItemGroup Condition="'$(PublishTrimmed)' != 'true'">
    <Compile Remove="generated\**" />
    <EmbeddedResource Remove="generated\**" />
    <None Remove="generated\**" />
  </ItemGroup>

  <ItemGroup>
    <ProjectReference Include="..\MyGenerator.csproj" 
      OutputItemType="Analyzer" 
      ReferenceOutputAssembly="false" 
      PrivateAssets="All" 
      Condition="'$(PublishTrimmed)' != 'true'" />
  </ItemGroup>
```

This works because the source generator has already generated the necessary files during the initial normal build phase, then they are available during the trimmed publish even without the generator reference.

**Comment by ritchiecarroll** [comment]
NOTE: This does **_not_** appear to be a duplicate of https://github.com/dotnet/runtime/issues/91535 - as this issue seems to be an issue of misidentification of source generation (analyzer) project (required to be .NET Standard 2.0) needing to be trimmed.

**Comment by jaredpar** [comment]
@agocke is this a trimmer issue? Can't see how this could be a compiler / generator issue.

**Comment by agocke** [comment]
Yeah, this appears to be something in the SDK related to publishing/trimming. @ritchiecarroll does this only repro in VS? Or does it also repro on the command line using `dotnet publish`?
